File: java/20. Valid Parentheses.java
// https://leetcode-cn.com/problems/valid-parentheses/submissions/
class Solution {
    public boolean isValid(String s) {
        if (s == null || s.length() <=1) return false;
        Deque q = new LinkedList<Character>();
        int i = 0;
        while (i < s.length()) {
            char c = s.charAt(i);
            switch(c) {
                case '(':
                case '[':
                case '{':
                q.push(c);break;
                case ')':{
                    Character top = (Character)q.poll();
                    if (top ==null || top !='(')
                        return false;
                }
                break;
                case ']':{
                    Character top = (Character)q.poll();
                    if (top ==null || top !='[')
                        return false;
                }
                break;
                case '}':{
                    Character top = (Character)q.poll();
                    if (top ==null || top !='{')
                        return false;
                }
                break;
            }
            i++;
        }
        return q.isEmpty();
    }
}

File: golang/20.easy.valid-parentheses/20.valid-parentheses.go
/*
 * @lc app=leetcode id=20 lang=golang
 *
 * [20] Valid Parentheses
 *
 * https://leetcode.com/problems/valid-parentheses/description/
 *
 * algorithms
 * Easy (39.85%)
 * Likes:    7224
 * Dislikes: 297
 * Total Accepted:    1.4M
 * Total Submissions: 3.4M
 * Testcase Example:  '"()"'
 *
 * Given a string s containing just the characters '(', ')', '{', '}', '[' and
 * ']', determine if the input string is valid.
 *
 * An input string is valid if:
 *
 *
 * Open brackets must be closed by the same type of brackets.
 * Open brackets must be closed in the correct order.
 *
 *
 *
 * Example 1:
 *
 *
 * Input: s = "()"
 * Output: true
 *
 *
 * Example 2:
 *
 *
 * Input: s = "()[]{}"
 * Output: true
 *
 *
 * Example 3:
 *
 *
 * Input: s = "(]"
 * Output: false
 *
 *
 * Example 4:
 *
 *
 * Input: s = "([)]"
 * Output: false
 *
 *
 * Example 5:
 *
 *
 * Input: s = "{[]}"
 * Output: true
 *
 *
 *
 * Constraints:
 *
 *
 * 1 <= s.length <= 10^4
 * s consists of parentheses only '()[]{}'.
 *
 *
 */

// @lc code=start

var pm = map[rune]rune{
	'(': ')',
	'{': '}',
	'[': ']',
}

func isValid(s string) bool {
	stack := make([]rune, 0, len(s))
	for _, r := range s {
		switch r {
		case '(', '{', '[':
			stack = append(stack, pm[r])
		case ')', '}', ']':
			if len(stack) == 0 {
				return false
			}
			if stack[len(stack)-1] != r {
				return false
			}
			stack = stack[:len(stack)-1]
		}
	}
	return len(stack) == 0
}

// @lc code=end

